File: analyse_nttestrecord.m
function record = analyse_nttestrecord(record,verbose)
%analyse_nttestrecord Analyses neurotar experiment
%
%  RECORD = analyse_nttestrecord( RECORD, [VERBOSE=true])
%
%  See nt_data_structures.md for more information.
%
% 2023-2025, Alexander Heimel, Zhiting Ren

if nargin<2 || isempty(verbose)
    verbose = true;
end

params = nt_default_parameters( record );

if ~isempty(params.nt_seed)
    rng(params.nt_seed);
end

nt_data = nt_load_tracking_data(record);

% [nt_data,neurotar_filename] = nt_load_neurotar_data(record);
% if isempty(nt_data)
%     nt_data = nt_load_mouse_tracks(record);
% end
% if isempty(nt_data)
%      nt_data = nt_load_DLC_data(record);
% end

if isempty(nt_data)
    logmsg(['Could not any position data for ' recordfilter(record)]);
end

logmsg(['Analyzing ' recordfilter(record)]);


if isempty(nt_data) && params.automatically_track_mouse 
    time_range = [];
     record = nt_track_mouse(record,time_range,verbose);
end

%% Check-out markers
if nt_check_markers( record, params, verbose ) == false
    return
end

record.measures.snippets_tbins = (-params.nt_photometry_pretime + params.nt_photometry_bin_width/2):params.nt_photometry_bin_width:(params.nt_photometry_posttime-params.nt_photometry_bin_width/2);

%% Photometry analysis
[record,photometry] = nt_analyse_photometry(record,nt_data,verbose);

snippets = [];
if ~isempty(photometry) 
    snippets = nt_make_photometry_snippets(photometry,record.measures,params);
end

%% Motion analysis
snippets = nt_make_motion_snippets(nt_data,record.measures,snippets,params);

filename = fullfile(nt_session_path(record),'nt_snippets.mat');
save(filename,'snippets');


%% Compute event measures
record.measures = nt_compute_event_measures(snippets,record.measures,params);


%% Object independent session measures
measures = record.measures;

if isempty(nt_data)
    logmsg('No nt_data to analyze.')
    return
end

mask = (nt_data.Time>0);
n_samples = sum(mask);

measures.session_speed_mean = mean(nt_data.Speed(mask),'omitnan');
measures.session_speed_std = std(nt_data.Speed(mask),'omitnan');
measures.session_speed_max = max(nt_data.Speed(mask));
measures.session_forward_speed_mean = mean(nt_data.Forward_speed(mask),'omitnan');
measures.session_forward_speed_std = std(nt_data.Forward_speed(mask),'omitnan');
measures.session_forward_speed_max = max(nt_data.Forward_speed(mask));
measures.session_angular_velocity_mean = mean(nt_data.Angular_velocity(mask),'omitnan');
measures.session_angular_velocity_std = std(nt_data.Angular_velocity(mask),'omitnan');
measures.session_angular_velocity_max = max(nt_data.Angular_velocity(mask));

measures.session_fraction_running_forward = sum( nt_data.Forward_speed(mask)> params.nt_min_approach_speed ) / n_samples;
measures.session_count_start_running_forward = sum( diff(nt_data.Forward_speed(mask) > params.nt_min_approach_speed )>0);
measures.session_start_running_forward_per_min = measures.session_count_start_running_forward / nt_data.Time(end) * 60;

measures.session_fraction_moving_backward = sum( nt_data.Forward_speed(mask) < params.nt_min_retreat_speed ) / n_samples;
measures.session_count_start_moving_backward = sum( diff(nt_data.Forward_speed(mask) < params.nt_min_retreat_speed )>0);
measures.session_start_moving_backward_per_min = measures.session_count_start_moving_backward / nt_data.Time(end) * 60;

record.measures = measures;

record = nt_compute_locations(record,nt_data,verbose);


if ~params.neurotar
    return
    % Below this only works for one object
end

%% Object analysis
[record,nt_data] = nt_object_analysis(record,nt_data,verbose);
measures = record.measures;
% if ~isempty(neurotar_filename)
%     save([neurotar_filename  '.mat'],'nt_data');
% end

%% Compute shuffles for object analysis
if isfield(measures,'object_positions') && ~isempty(measures.object_positions)

    for b = 1:length(params.nt_behaviors)
        behavior = params.nt_behaviors(b).behavior;
        measures.(behavior).shuffles_count_in_interaction_period = [];
        measures.(behavior).shuffles_fraction_in_interaction_period = [];
    end % b
    for i = 1:length(params.nt_indices)
        index = params.nt_indices(i).index;
        measures.(index).shuffles = [];
    end % i
    for r = 1:length(params.nt_rates)
        rate = params.nt_rates(r).rate;
        measures.(rate).shuffles_time = [];
        measures.(rate).shuffles_mean = [];
    end % r

    % Shuffle object insertions
    for i = 1:params.nt_shuffle_number
        logmsg(['Shuffle ' num2str(i) ' of ' num2str(params.nt_shuffle_number)])
        shuffle_record = shuffle_object_insertions( record,nt_data,params );
        shuffle_record = nt_object_analysis(shuffle_record,nt_data,false);

        for b = 1:length(params.nt_behaviors)
            behavior = params.nt_behaviors(b).behavior;
            measures.(behavior).shuffles_count_in_interaction_period = ...
                [measures.(behavior).shuffles_count_in_interaction_period shuffle_record.measures.(behavior).count_in_interaction_period];
            measures.(behavior).shuffles_fraction_in_interaction_period = ...
                [measures.(behavior).shuffles_fraction_in_interaction_period shuffle_record.measures.(behavior).fraction_in_interaction_period];
        end % b

        for j = 1:length(params.nt_indices)
            index = params.nt_indices(j).index;
            measures.(index).shuffles(i) = shuffle_record.measures.(index).val;
        end % j

        for r = 1:length(params.nt_rates)
            rate = params.nt_rates(r).rate;
            measures.(rate).shuffles_time = [measures.(rate).shuffles_time shuffle_record.measures.(rate).time];
            measures.(rate).shuffles_mean = [measures.(rate).shuffles_mean shuffle_record.measures.(rate).mean];
        end % r
    end % shuffle i

    % compute mean rates for shuffles
    for r = 1:length(params.nt_rates)
        rate = params.nt_rates(r).rate;
        [~,ind] = sort(measures.(rate).shuffles_time);
        [measures.(rate).shuffles_mean,measures.(rate).shuffles_time,measures.(rate).shuffles_sem] = ...
            slidingwindowfunc(measures.(rate).shuffles_time(ind), measures.(rate).shuffles_mean(ind), ...
            -3, 0.1, params.nt_interaction_period,params.nt_object_sliding_window,'nanmean');
    end % r
else
    logmsg('No object positions given. Retrack?');
end
record.measures = measures;

end

%% Help functions

function shuffle_record = shuffle_object_insertions( record, neurotar_data, params )
% Shuffles object markers and positions into new record
shuffle_record = record;

if ~params.neurotar
    logmsg('SHUFFLE STILL NEEDS TO BE UPDATED FOR FREELY WALKING. DISABLED')
    return
end

neurotar_duration = neurotar_data.Time(end);
measures = record.measures;

ind = find([record.measures.markers.marker]~='o' & ...
    [record.measures.markers.marker]~='t' & ...
    [record.measures.markers.marker]~='v' & ...
    [record.measures.markers.marker]~='h' & ...
    [record.measures.markers.marker]~='f');
measures.markers(ind) = [];
if measures.markers(end).marker~=params.nt_stop_marker
    % Add stop marker at the end
    measures.markers(end+1).marker = params.nt_stop_marker;
    measures.markers(end).time = neurotar_data.Time(end);
end
shuffle_record.measures = measures;


if params.nt_shuffle_insert_object_only_when_stationary

    if 0 % new method
        stationary = abs(neurotar_data.Forward_speed < params.nt_maximum_stationarity_speed);
        stationary(1) = false; % set moving at start
        stationary(end) = false; % set moving at start
        start_stationary = find(diff(stationary)>0);
        end_stationary = find(diff(stationary)<0);
        stationary_period = neurotar_data.Time(end_stationary) - neurotar_data.Time(start_stationary);
        ind_suitable_periods = find(stationary_period > params.nt_shuffle_stationary_period);
        if isempty(ind_suitable_periods)
            errormsg('Too few long enough stationary periods for shuffling. Consider lowering params.nt_shuffle_stationary_period');
            return
        end

        logmsg('NOT FINISHED IMPLEMENTING NEW METHOD YET')

    else % old method
        random_timeshift = 0;

        m = 1;
        while m<=length(measures.markers)
            if contains('ovhf', measures.markers(m).marker)
                moving = true;
                splitup = true;
                while moving || splitup
                    random_timeshift = neurotar_duration*rand(1);
                    shuffle_record.measures.markers(m).time = mod(measures.markers(m).time + random_timeshift,neurotar_duration);
                    ind = find(neurotar_data.Time>shuffle_record.measures.markers(m).time-params.nt_shuffle_stationary_period & neurotar_data.Time<shuffle_record.measures.markers(m).time );
                    moving = any(abs(neurotar_data.Forward_speed(ind)) > params.nt_max_stationarity_speed);
                    ind_take_out = m + find([measures.markers(m+1:end).marker]=='t',1);
                    splitup = (mod(measures.markers(ind_take_out).time + random_timeshift,neurotar_duration) < shuffle_record.measures.markers(m).time);
                end
            else
                shuffle_record.measures.markers(m).time = mod(measures.markers(m).time + random_timeshift,neurotar_duration);
            end

            % shift object positions
            if m<length(measures.markers)
                ind = find( measures.object_positions(:,1) > measures.markers(m).time & measures.object_positions(:,1) < measures.markers(m+1).time);
            else
                ind = find( measures.object_positions(:,1) > measures.markers(m).time);
            end
            for j = ind(:)'
                shuffle_record.measures.object_positions(j,1) = mod(shuffle_record.measures.object_positions(j,1) + random_timeshift,neurotar_duration);
                shuffle_record.measures.object_positions(j,[2 3]) = shuffle_record.measures.object_positions(j,[3 2]);  % swap x and y
            end

            m = m + 1;
        end
    end
else
    random_timeshift = neurotar_duration*rand(1);
    for m = 1:length(measures.markers)
        shuffle_record.measures.markers(m).time = mod(measures.markers(m).time + random_timeshift,neurotar_duration);
    end

    % shift object positions accordingly
    shuffle_record.measures.object_positions(:,1) = mod(record.measures.object_positions(:,1) + random_timeshift,neurotar_duration);
    shuffle_record.measures.object_positions(:,[2 3]) = record.measures.object_positions(:,[3 2]);  % swap x and y

end

% resort markers
[~,ind] = sort([shuffle_record.measures.markers(:).time]);
shuffle_record.measures.markers = shuffle_record.measures.markers(ind);

% resort object positions
[~,ind] = sort(shuffle_record.measures.object_positions(:,1));
shuffle_record.measures.object_positions = shuffle_record.measures.object_positions(ind,:);


end

File: novitrack/analyse_nttestrecord.py
"""Top-level NoviTrack session analysis.

This is a first Python port of the non-GUI, already-tracked path through
``analyse_nttestrecord.m``. Neurotar object analysis and tracking recompute
branches are intentionally left for later.
"""

from __future__ import annotations

from collections.abc import Mapping
from copy import deepcopy
from pathlib import Path
from typing import Any

import numpy as np
from scipy.io import savemat

from inpythotools.logmsg import logmsg
from .nt_add_surgery_info import nt_add_surgery_info
from .nt_analyse_photometry import nt_analyse_photometry
from .nt_check_markers import nt_check_markers
from .nt_compute_event_measures import nt_compute_event_measures
from .nt_compute_locations import nt_compute_locations
from .nt_load_parameters import nt_load_parameters
from .nt_load_tracking_data import nt_load_tracking_data
from .nt_make_motion_snippets import nt_make_motion_snippets
from .nt_make_photometry_snippets import nt_make_photometry_snippets
from .nt_session_path import nt_session_path


def _get(obj: Any, name: str, default: Any = None) -> Any:
    if obj is None:
        return default
    if isinstance(obj, Mapping):
        return obj.get(name, default)
    return getattr(obj, name, default)


def _as_array(value: Any) -> np.ndarray:
    return np.asarray(value, dtype=float).reshape(-1)


def _record_label(record: Mapping[str, Any]) -> str:
    return str(_get(record, "sessionid", _get(record, "subject", "record")))


def _session_measures(measures: dict[str, Any], nt_data: Mapping[str, Any], params: Any) -> dict[str, Any]:
    time = _as_array(_get(nt_data, "Time"))
    mask = time > 0
    n_samples = int(np.sum(mask))
    if n_samples == 0:
        return measures

    speed = _as_array(_get(nt_data, "Speed"))
    forward_speed = _as_array(_get(nt_data, "Forward_speed"))
    angular_velocity = _as_array(_get(nt_data, "Angular_velocity"))

    def nanmean(values: np.ndarray) -> float:
        values = values[np.isfinite(values)]
        return float(np.mean(values)) if values.size else np.nan

    def nanstd(values: np.ndarray) -> float:
        values = values[np.isfinite(values)]
        return float(np.std(values, ddof=1)) if values.size > 1 else np.nan

    def nanmax(values: np.ndarray) -> float:
        values = values[np.isfinite(values)]
        return float(np.max(values)) if values.size else np.nan

    measures["session_speed_mean"] = nanmean(speed[mask])
    measures["session_speed_std"] = nanstd(speed[mask])
    measures["session_speed_max"] = nanmax(speed[mask])
    measures["session_forward_speed_mean"] = nanmean(forward_speed[mask])
    measures["session_forward_speed_std"] = nanstd(forward_speed[mask])
    measures["session_forward_speed_max"] = nanmax(forward_speed[mask])
    measures["session_angular_velocity_mean"] = nanmean(angular_velocity[mask])
    measures["session_angular_velocity_std"] = nanstd(angular_velocity[mask])
    measures["session_angular_velocity_max"] = nanmax(angular_velocity[mask])

    min_approach_speed = float(_get(params, "nt_min_approach_speed"))
    min_retreat_speed = float(_get(params, "nt_min_retreat_speed"))
    forward_masked = forward_speed[mask]

    measures["session_fraction_running_forward"] = float(np.sum(forward_masked > min_approach_speed) / n_samples)
    measures["session_count_start_running_forward"] = int(np.sum(np.diff(forward_masked > min_approach_speed) > 0))
    measures["session_start_running_forward_per_min"] = float(
        measures["session_count_start_running_forward"] / time[-1] * 60
    )
    measures["session_fraction_moving_backward"] = float(np.sum(forward_masked < min_retreat_speed) / n_samples)
    measures["session_count_start_moving_backward"] = int(np.sum(np.diff(forward_masked < min_retreat_speed) > 0))
    measures["session_start_moving_backward_per_min"] = float(
        measures["session_count_start_moving_backward"] / time[-1] * 60
    )

    return measures


def analyse_nttestrecord(
    record: Mapping[str, Any],
    *,
    params: Any | None = None,
    yaml_file: str | Path | None = None,
    session_path: str | Path | None = None,
    photometry_folder: str | Path | None = None,
    save_snippets: bool | None = None,
    verbose: bool = True,
) -> dict[str, Any]:
    """Analyze one NoviTrack record and return an updated record dictionary."""
    out = deepcopy(dict(record))
    out.setdefault("measures", {})

    if params is None:
        params = nt_load_parameters(out, yaml_file=yaml_file)

    if save_snippets is None:
        save_snippets = bool(_get(params, "nt_save_snippets", True))

    if _get(params, "nt_seed", None):
        np.random.seed(int(_get(params, "nt_seed")))

    out = nt_add_surgery_info(out, params)

    nt_data, trigger_times = nt_load_tracking_data(out, params, recompute=False, session_path=session_path)
    if not nt_data:
        logmsg(f"Could not find any position data for {_record_label(out)}")

    if verbose:
        logmsg(f"Analyzing {_record_label(out)}")

    measures = dict(_get(out, "measures", {}))
    measures["event"] = {}
    measures.pop("events", None)
    if trigger_times.size:
        measures["trigger_times"] = trigger_times

    if "markers" not in measures or len(measures.get("markers", [])) == 0:
        measures["markers"] = []

    out["measures"] = measures
    if not nt_check_markers(out, params, verbose=verbose):
        return out

    bin_width = float(_get(params, "nt_photometry_bin_width"))
    pretime = float(_get(params, "nt_pretime"))
    posttime = float(_get(params, "nt_posttime"))
    measures["snippets_tbins"] = np.arange(
        -pretime + bin_width / 2,
        posttime - bin_width / 2 + bin_width / 10,
        bin_width,
    )
    out["measures"] = measures

    out, photometry, _ = nt_analyse_photometry(
        out, nt_data, params, photometry_folder=photometry_folder
    )
    measures = dict(out["measures"])

    snippets: dict[str, Any] = {}
    if photometry:
        snippets = nt_make_photometry_snippets(photometry, measures, params)

    snippets = nt_make_motion_snippets(nt_data, measures, snippets, params)
    if save_snippets:
        if session_path is None:
            folder, exists = nt_session_path(out, params)
        else:
            folder = Path(session_path)
            exists = folder.is_dir()
        if exists:
            filename = folder / "nt_snippets.mat"
            try:
                savemat(str(filename), {"snippets": snippets}, do_compression=True, long_field_names=True)
            except (OSError, TypeError, ValueError) as exc:
                logmsg(f"Could not save snippets to {filename}: {exc}")

    measures = nt_compute_event_measures(snippets, measures, params)

    if nt_data:
        measures = _session_measures(measures, nt_data, params)
        out["measures"] = measures
        if not bool(_get(params, "neurotar", False)):
            out = nt_compute_locations(out, nt_data, params)
        else:
            logmsg("Compute locations is not yet implemented for neurotar.")
    else:
        out["measures"] = measures

    if bool(_get(params, "neurotar", False)):
        logmsg("Neurotar object analysis and shuffling are not ported yet.")

    # Keep snippets available to callers without storing them inside measures.
    out["snippets"] = snippets
    return out
